sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,233 @@
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Represents a signal edge type.
+ */
+typedef enum dif_spi_device_edge {
+  kDifSpiDeviceEdgePositive,
+  kDifSpiDeviceEdgeNegative,
+} dif_spi_device_edge_t;
+
+/**
+ * Represents a signal edge type.
+ */
+typedef enum dif_spi_device_bit_order {
+  kDifSpiDeviceBitOrderMsbToLsb,
+  kDifSpiDeviceBitOrderLsbToMsb,
+} dif_spi_device_bit_order_t;
+
+typedef enum dif_spi_device_irq_state {
+  kDifSpiDeviceIrqStateEnable = true,
+  kDifSpiDeviceIrqStateDisable = false,
+} dif_spi_device_irq_state_t;
+
+/**
+ * Represents the types of interrupts that the SPI device will
+ * fire to signal state.
+ */
+typedef enum dif_spi_device_irq_type {
+  kDifSpiDeviceIrqTypeRxFull,
+  kDifSpiDeviceIrqTypeRxAboveLevel,
+  kDifSpiDeviceIrqTypeTxBelowLevel,
+  kDifSpiDeviceIrqTypeRxError,
+  kDifSpiDeviceIrqTypeRxOverflow,
+  kDifSpiDeviceIrqTypeTxUnderflow,
+} dif_spi_device_irq_type_t;
+
+/**
+ * Errors returned by SPI DIF functions.
+ */
+typedef enum dif_spi_device_result {
+  /**
+   * Indicates a successful operation.
+   */
+  kDifSpiDeviceResultOk = 0,
+  /**
+   * Indicates a failed precondition on the function arguments.
+   */
+  kDifSpiDeviceResultBadArg,
+} dif_spi_device_result_t;
+
+/**
+ * The length of the SPI device SRAM, in bytes.
+ *
+ * Useful for initializing FIFO lengths.
+ */
+extern const uint16_t kDifSpiDeviceSramLen;
+
+/**
+ * Configuration for initializing a SPI device.
+ */
+typedef struct dif_spi_device_config {
+  mmio_region_t base_addr;
+  dif_spi_device_edge_t clock_polarity;
+  dif_spi_device_edge_t data_phase;
+  dif_spi_device_bit_order_t tx_order;
+  dif_spi_device_bit_order_t rx_order;
+  uint16_t rx_fifo_timeout;  // uint12_t
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_len;
+} dif_spi_device_config_t;
+
+/**
+ * State for a particular SPI device.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_spi_device {
+  mmio_region_t base_addr;
+  uint16_t rx_fifo_base;
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_base;
+  uint16_t tx_fifo_len;
+} dif_spi_device_t;
+
+/**
+ * Initializes a SPI device with the given configuration.
+ *
+ * @param config the configuration to initialize with.
+ * @param spi out param for the initialized device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_init(dif_spi_device_config_t config,
+                                            dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ *
+ * @param spi a SPI device.
```

Comment:
Redundant `copy/paste`, already declared above.

Reply:
Yep, got it.